### PR TITLE
Release process with Docker image published to ghcr.io

### PR DESCRIPTION
**Pull request by wallrj**
A release process  which is triggered by a developer filling in the GitHub `Create New Release` page.
I considered something which creates the entire GitHub release automatically when ever a new tag is pushed, but this seemed like a good first step, and it think it is  intuitive that a developer can trigger a release using the GitHub UI.

Fixes: #3 

You can test this by creating a pre-release yourself, but based on this `3-publish-docker-image` branch rather than `main` branch.

image

Then visit the `Actions` tab to see the automated release logs:

image

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
     container: golang:1.16
     steps:
-      - name: Checkout
+      -
+        name: Checkout
         uses: actions/checkout@v2
-      - name: Run tests
+      -
+        name: Run tests
         run: make test
```
Comment (wallrj) [inline .github/workflows/test.yaml:13]: These changes are unrelated to the release process, but it fixes a bug where the tests were being run twice, triggered by both the `pull_request` event and by the `push` event. GitHub actions doesn't seem to de-duplicate the events.

File: README.md [hunk 2]
```diff
@@ -2,14 +2,37 @@
 
 This is an example of an [External Issuer] for cert-manager.
 
+## Install
+
+```
+kubectl apply -f https://github.com/cert-manager/sample-external-issuer/releases/download/v0.1.0/install.yaml
+```
```
Comment (wallrj) [inline README.md:9]: Once this merges, I'll do a `v0.1.0` release, and this instruction will be correct.

File: README.md [hunk 3]
```diff
@@ -2,14 +2,37 @@
...
+
 ## Demo
 
 You can run the sample-external-issuer on a local cluster with this command:
 
 ```
-make IMG=controller:0.0.0 kind-cluster deploy-cert-manager docker-build kind-load deploy e2e
+make kind-cluster deploy-cert-manager docker-build kind-load deploy e2e
```
Comment (wallrj) [inline README.md:16]: The IMG value is derived from the `git describe --tags` output, so it's no longer necessary to specify it here.

File: README.md [hunk 4]
```diff
@@ -2,14 +2,37 @@
...
+ * Target: `main`
+ * Release Title: `Release v0.1.0-alpha.2`
+ * Description: (optional) a short summery of the changes since the last release.
+
+Click the `Publish release` button to trigger the automated release process:
+
+* A Docker image will be generated and published to `gcr.io/cert-manager/sample-external-issuer/controller` with the chosen tag.
+* An `install.yaml` file will be generated and attached to the release.
+
+[GitHub New Release Page]: https://github.com/cert-manager/sample-external-issuer/releases/new
```
Comment (wallrj) [inline README.md:34]: This is very basic, but I hope it is sufficient for other devs to follow.

File: Makefile [hunk 5]
```diff
@@ -54,12 +58,23 @@ install: manifests
 uninstall: manifests
 	kustomize build config/crd | kubectl delete -f -
 
-# Deploy controller in the configured Kubernetes cluster in ~/.kube/config
-deploy:
-	pushd config/manager
+# TODO(wallrj): .PHONY ensures that the install file is always regenerated,
+# because I this really depends on the checksum of the Docker image and all the
```
Comment (jakexks) [inline Makefile:62]: Does this mean you want to ensure the docker image doesn't contain anything stale? I can't make sense of this sentence

**Review by jakexks — APPROVED** [review]
Looks good. I prefer the pushing tag creates a draft release as it ensures artifacts are built before releasing, but I can see the arguments for either way around.

File: Makefile (hunk 5 above)
Comment (wallrj) [reply Makefile:62]: I meant that without `.PHONY` the `build/install.yaml` file will be built once and never regenerated when either the kustomize `config/` files change *OR* when the Docker image version number changes.

I could create a Make rule for all those files and all the `.go` files, using `find`, I suppose.
But this seemed easier, for now.

**Comment by wallrj** [comment]
Thanks @jakexks 

> Looks good. I prefer the pushing tag creates a draft release as it ensures artifacts are built before releasing, but I can see the arguments for either way around.

I like that idea too and I might create a followup PR where I trigger the release that way.
Merging this now.

/lgtm
/approve

**Pull request merged**